### PR TITLE
Include tokens from function output in streaming callbacks

### PR DESCRIPTION
**Pull request by dmyoung9**
Description
---
This adds a small bit of logic so that `on_new_llm_token` callbacks will be properly passed a valid token (instead of `''`) when a streaming model is called as part of a function call or tool run.

Issue
---
This is a *partial* fix for #7385, having only implemented this logic for `ChatOpenAI`. As @rasca mentions in that issue:

> We need to add this to every llm/

I haven't gone this far, as I don't have reliable (or any) access to many of the LLMs that are integrated, but am using OpenAI's models for my own development.

Dependencies
---
There are no extra dependencies required for this fix.

Tag Maintainer
---
I have a feeling this one should probably be looked at by @agola11.

Twitter Handle
---
What's Twitter? 🤣

### CODE REVIEW

**Comment by dmyoung9** [comment]
Looks like a formatting issue, but should be an easy fix.

**Comment by dmyoung9** [comment]
I've fixed the linter error, and added the same logic to the regular `OpenAI` LLM as well.

**Comment by dmyoung9** [comment]
I've fixed the linter error, but had made a mistake, so I force pushed 😬

In addition, I'm wondering if this looks to be the right way to implement this same change in `BaseOpenAI`, before I push that as well 😅:
```diff
diff --git a/libs/langchain/langchain/llms/openai.py b/libs/langchain/langchain/llms/openai.py
index dacafe0f..372ac08c 100644
--- a/libs/langchain/langchain/llms/openai.py
+++ b/libs/langchain/langchain/llms/openai.py
@@ -296,8 +296,14 @@ class BaseOpenAI(BaseLLM):
             chunk = _stream_response_to_generation_chunk(stream_resp)
             yield chunk
             if run_manager:
+                chunk_content = chunk.additional_kwargs.get("function_call", None)
+                token = (
+                    chunk_content["arguments"]
+                    if chunk_content is not None
+                    else chunk.text
+                )
                 run_manager.on_llm_new_token(
-                    chunk.text,
+                    token,
                     verbose=self.verbose,
                     logprobs=chunk.generation_info["logprobs"]
                     if chunk.generation_info
@@ -319,8 +325,14 @@ class BaseOpenAI(BaseLLM):
             chunk = _stream_response_to_generation_chunk(stream_resp)
             yield chunk
             if run_manager:
+                chunk_content = chunk.additional_kwargs.get("function_call", None)
+                token = (
+                    chunk_content["arguments"]
+                    if chunk_content is not None
+                    else chunk.text
+                )
                 await run_manager.on_llm_new_token(
-                    chunk.text,
+                    token,
                     verbose=self.verbose,
                     logprobs=chunk.generation_info["logprobs"]
                     if chunk.generation_info
```

**Comment by dmyoung9** [comment]
I've also put together a simple script to illustrate the issue at hand here:

```python
# function_call_gets_no_tokens.py

from dotenv import load_dotenv

load_dotenv()

from typing import Any
from uuid import UUID
from langchain.callbacks.base import BaseCallbackHandler
from langchain.chains import LLMChain
from langchain.chains.openai_functions import create_structured_output_chain
from langchain.chat_models import ChatOpenAI
from langchain.output_parsers import PydanticOutputParser
from langchain.prompts import PromptTemplate

from pydantic import BaseModel, Field


class CustomCallbackHandler(BaseCallbackHandler):
    def on_llm_new_token(
        self,
        token: str,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> Any:
        print(f"'{token}'")  # wrap token in single quotes for visibility

        return super().on_llm_new_token(
            token, run_id=run_id, parent_run_id=parent_run_id, **kwargs
        )


class Joke(BaseModel):
    setup: str = Field(description="question to set up a joke")
    punchline: str = Field(description="answer to resolve the joke")


chat_model = ChatOpenAI(
    streaming=True,
    callbacks=[CustomCallbackHandler()],
)

joke_parser = PydanticOutputParser(pydantic_object=Joke)
prompt_template = "Tell me a joke about {subject}.\n\n{format_instructions}"
chat_prompt = PromptTemplate(
    template=prompt_template,
    input_variables=["subject"],
    partial_variables={"format_instructions": joke_parser.get_format_instructions()},
)


def chain_as_normal_call(subject):
    chain = LLMChain(llm=chat_model, prompt=chat_prompt)
    return chain(subject)


def chain_as_function_call(subject):
    chain = create_structured_output_chain(Joke, chat_model, chat_prompt)
    return chain(subject)


print("Tokens received as part of a normal call:")
normal_joke = chain_as_normal_call("frogs")
print(normal_joke)
print()

print("Tokens not received as part of a function call:")
function_joke = chain_as_function_call("frogs")
print(function_joke)
print()
```

Anyone should be able to simply run that script and see the issue, but here is a sample output:
```
Tokens received as part of a normal call:
''
'{"'
'setup'
'":'
' "'
'Why'
' was'
' the'
' frog'
' happy'
'?",'
' "'
'p'
'unch'
'line'
'":'
' "'
'Because'
' he'
' ate'
' a'
' fly'
'!"'
'}'
''
{'subject': 'frogs', 'text': '{"setup": "Why was the frog happy?", "punchline": "Because he ate a fly!"}'}

Tokens not received as part of a function call:
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
''
{'subject': 'frogs', 'function': Joke(setup='Why are frogs so happy?', punchline='Because they eat whatever bugs them!')}
```

**Comment by hinthornw** [comment]
Thanks for putting this together, @dmyoung9 ! Tagging in @baskaryan as well. I'll try to review later today if not

File: libs/langchain/langchain/chat_models/openai.py [hunk 1]
```diff
@@ -374,7 +374,13 @@ def _stream(
             default_chunk_class = chunk.__class__
             yield ChatGenerationChunk(message=chunk)
             if run_manager:
-                run_manager.on_llm_new_token(chunk.content)
+                chunk_content = chunk.additional_kwargs.get("function_call", None)
+                token = (
+                    chunk_content["arguments"]
+                    if chunk_content is not None
+                    else chunk.content
```
Comment (hinthornw) [inline libs/langchain/langchain/chat_models/openai.py:381]: What if the chunk has both content and arguments?

Comment (dmyoung9) [reply libs/langchain/langchain/chat_models/openai.py:381]: From what I've seen in testing, this does not happen. When the chain is called as part of a function/tool call, the output is in the form of a function, and the "standard" chunk content is empty.

Comment (dmyoung9) [reply libs/langchain/langchain/chat_models/openai.py:381]: @AI-Chef To be fair, that is probably a safer way of handling this for the base class as well! I'll push a commit to fix that in a little bit 👌

**Comment by hinthornw** [comment]
@dmyoung9 we recently added the full "chunk" to `on_llm_new_token` callbacks. I think that would resolve this issue in a slightly cleaner way

File: libs/langchain/langchain/chat_models/openai.py (hunk 1 above)
Comment (hinthornw) [reply libs/langchain/langchain/chat_models/openai.py:322]: The funciton calling endpoint does sometimes return both, and in any case it wouldn't be a behavior we could bank on long-term, so we added the whole chunk to the callback

**Comment by dmyoung9** [comment]
@hinthornw I'll check it out in my end, but that does seem like it would also work. I'll close if that works for me.

**Comment by hinthornw** [comment]
Thank you for pushing on this!

**Comment by leo-gan** [comment]
@dmyoung9 Hi , could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!
If this PR is not needed anymore, could you, please, let me know or close this PR?

**Comment by efriis** [comment]
@dmyoung9 Looks like this isn't being worked on anymore. I'll close for now, and if you resume work, feel free to reopen!

**Pull request closed without merging**